### ISSUE TITLE
[APP CRASH] int 2d not handled correctly

### ISSUE THREAD

**Issue by dcdelia** [DynamoRIO/dynamorio#4569]
**Describe the bug**
DynamoRIO does not adjust EIP by +1 byte as expected on Windows for an `int 2d` instruction in an application with custom SEH debugger-like code, but execution rather gets stuck and the program never completes.

**To Reproduce**
I have attached both ASM source and EXE (compiled with MASM32 with Console target) of the original [debugger detection](http://www.openrce.org/reference_library/anti_reversing_view/34/INT%202D%20Debugger%20Detection/) by Rolf Rolles from OpenRCE and a minified version that I obtained for it. To run the program just use `drrun.exe -- int2d-min.exe`. The program does not get to execute MessageBox.

[int2d.zip](https://github.com/DynamoRIO/dynamorio/files/5619455/int2d.zip)

**Expected behavior**
The program should skip the `inc esi` instruction after `int 2d` and print in a MessageBox that ESI is 0.

```
start:

assume fs:nothing
mov esi, 0
push offset _seh ;\ see ASM file for handler
push fs:[0] ; > set SEH
mov fs:[0], esp ;/
int 2dh ; 
inc esi ; 1 byte or it will jump halfway in pop fs:[0]

pop fs:[0] ;\ clear SEH
add esp, 4 ;/

push esi
[...] ; part that turns it into a wchar_t
call MessageBox
```

**Versions**
I have used v.8.0.0 build 1 as well as the latest build 18585. 32-bit application executed on Windows 7 SP1 x64.

**Additional context**
I am poking DynamoRIO for DBI detection methods research and saw that fixes done in the past (e.g. #2218 and #2250) work with other variants of `int 2d` that I have been trying (I can share them if useful). Attached also a log for `drrun.exe -verbose -s 30 -debug -loglevel 4`, couldn't see information related to the problem in it.

**Comment by derekbruening** [comment]
Thank you for the filing.  Just for my own summary: so the PR #2218 that you've linked, for #2126, added a test and handling for `int 2d` raising an exception where ExceptionAddress is 1 byte too far (yet what counts, the CONTEXT.Eip, is correct); while #2250 covers the cases where it does not raise an exception (even when not under a debugger), but has not been implemented yet.

The problem in #2126 was just DR itself complaining about ExceptionAddress not being on an instruction boundary.  I do not think there was any actual behavior problem in the app.

So what is going on here and why is it different?  Is this SEH handler trying to target ExceptionAddress instead of CONTEXT.Eip?  DR has the translated versions matching the original, so we would expect either to match native behavior.

**Comment by dcdelia** [comment]
The handler is a bit extreme as it sets up a mini-debugger:

```
_seh:
push offset procinfo
push offset startinfo
push 0
push 0
push DEBUG_PROCESS
push 0
push 0
push 0
call GetCommandLine
push eax
push 0
call CreateProcess

_dbgloop:
push INFINITE
push offset debugEvt
call WaitForDebugEvent

cmp debugEvt.dwDebugEventCode, EXIT_PROCESS_DEBUG_EVENT
je _end

push DBG_CONTINUE
push debugEvt.dwThreadId
push debugEvt.dwProcessId
call ContinueDebugEvent

jmp _dbgloop

_end: push 0
call ExitProcess
```

If I make `_seh` call ExitProcess(1) immediately without the CreateProcess & WaitForDebugEvent part DynamoRIO works as expected. In my tests it also works fine with standard anti-debug patterns where `_seh` is a block that checks EIP offset 0xb8 in the Context against the expected instruction pointer and jumps back into ordinary code, so I would say the PR and fixes you quoted above did their job.

**Comment by derekbruening** [comment]
So the first invocation turns itself into a debugger, launching a new process with the same command line with itself attached as the debugger.  Then the child's `int 2d` raises the breakpoint exception, and the parent says `DBG_CONTINUE`, which marks the exception as handled, which is why the child does not also try to launch a new process.

I think this is similar to #2389: DR in the child has no chance to do anything, including translate the context and ExceptionAddress: the kernel handed the raw code cache context to the debugger and the debugger says "go here".  If that target is the ExceptionAddress, which is the pc+1, it might target the middle of an instruction in the code cache or sthg.  Is there anything in the DR log after this point?

If that is what's happening, it's not clear what DR can do about it: arrange for the +1 in the cache to be something sane I guess.

**Comment by dcdelia** [comment]
I agree for the similarities with #2389. Perhaps arranging for the +1 could be a good workaround. The log says that both the parent and the child get to execute int2d, then no more output until the timeout message. Excerpt:
```<dynamo auto start>
<non-syscall, non-int2b 0x2d @ 0x00401018 from 0x00401000>
<found a fork: pid 2436>
<log dir=C:\Documents and Settings\pctips\Desktop\drio801\logs\int2d-min.exe.2436.00000000>
<Starting application C:\Documents and Settings\pctips\Desktop\dbi-bugs\PoCs-DRio\int2d\int2d-min.exe (2436)>
<Initial options = -loglevel 4 -code_api -probe_api -stack_size 56K -max_elide_jmp 0 -max_elide_call 0 -no_inline_ignored_syscalls -native_exec_default_list '' -no_native_exec_managed_code -no_indcall2direct -no_aslr_dr >
<dynamo auto start>
<non-syscall, non-int2b 0x2d @ 0x00401018 from 0x00401000>
```

Both processes keep eating a few million CPU cycles until the timeout kicks in. The parent uses ~40% more cycles, writes twice as many bytes, and open 60 handles instead of 40, including a DebugObject one (I got this information with Process Hacker). I'd guess some looping going on, but I'm not confident yet with DynamoRIO to speculate whether it's a loop in the code cache or some "livelock" in the runtime.
